File: verifier-kotlin/src/main/kotlin/tech/thessemaj/deviceintelligence/verifier/ChainVerifier.kt
package tech.thessemaj.deviceintelligence.verifier

import java.security.MessageDigest
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate

/**
 * Validates the token's attestation cert chain: each cert signed by the next, and
 * the top signed by (or equal to) a pinned Google root. Port of
 * verify_token.py `_verify_chain_to_pinned` / `_verify_signed_by`.
 */
internal class ChainVerifier(private val pinnedRoots: List<X509Certificate>) {

    fun parseChain(certsHex: List<String>): List<X509Certificate> {
        val cf = CertificateFactory.getInstance("X.509")
        return certsHex.map { hex ->
            cf.generateCertificate(Hex.decode(hex).inputStream()) as X509Certificate
        }
    }

    /** Throws if [cert] is not signed by [issuer]'s public key. */
    private fun verifySignedBy(cert: X509Certificate, issuer: X509Certificate) {
        cert.verify(issuer.publicKey)   // signature only — validity dates are not our gate
    }

    private fun sha256Fp(cert: X509Certificate): ByteArray =
        MessageDigest.getInstance("SHA-256").digest(cert.encoded)

    /** Returns the pinned root the chain terminates in, or throws. */
    fun verifyToPinnedRoot(chain: List<X509Certificate>): X509Certificate {
        require(chain.isNotEmpty()) { "empty chain" }
        for (i in 0 until chain.size - 1) verifySignedBy(chain[i], chain[i + 1])
        val top = chain.last()
        val topFp = sha256Fp(top)
        for (root in pinnedRoots) {
            if (topFp.contentEquals(sha256Fp(root))) return root
            try {
                verifySignedBy(top, root)
                return root
            } catch (_: Exception) { /* try next root */ }
        }
        throw IllegalArgumentException("chain top does not chain to a pinned Google root")
    }
}

File: verifiers/node/src/chain_verifier.ts
import { createHash, verify as cryptoVerify, createPublicKey } from "node:crypto";
import { X509Certificate } from "node:crypto";

export function parseChain(certsHex: string[]): X509Certificate[] {
  return certsHex.map(h => new X509Certificate(Buffer.from(h, "hex")));
}

function sha256Fp(cert: X509Certificate): Buffer {
  return createHash("sha256").update(cert.raw).digest();
}

export function verifyToPinnedRoot(chain: X509Certificate[], pinnedRoots: X509Certificate[]): X509Certificate {
  if (!chain.length) throw new Error("empty chain");
  for (let i = 0; i < chain.length - 1; i++) verifySignedBy(chain[i], chain[i + 1]);
  const top = chain[chain.length - 1];
  const topFp = sha256Fp(top);
  for (const root of pinnedRoots) {
    if (topFp.equals(sha256Fp(root))) return root;
    try { verifySignedBy(top, root); return root; } catch { /* try next */ }
  }
  throw new Error("chain top does not chain to a pinned Google root");
}

function verifySignedBy(cert: X509Certificate, issuer: X509Certificate): void {
  if (!cert.verify(issuer.publicKey)) throw new Error("signature mismatch");
}
